pkg/pillar/cmd/usbmanager/usbmanager.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package usbmanager
+
+import (
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/agentbase"
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/cmd/domainmgr"
+	"github.com/lf-edge/eve/pkg/pillar/hypervisor"
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+	"github.com/lf-edge/eve/pkg/pillar/utils"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	agentName = "usbmanager"
+	// Time limits for event loop handlers
+	errorTime            = 3 * time.Minute
+	warningTime          = 40 * time.Second
+	stillRunningInterval = 25 * time.Second
+)
+
+var (
+	logger *logrus.Logger
+	log    *base.LogObject
+)
+
+type usbdevice struct {
+	busnum                  uint16
+	portnum                 string
+	devnum                  uint16
+	vendorID                uint32
+	productID               uint32
+	devicetype              string
+	usbControllerPCIAddress string
+	ueventFilePath          string
+}
+
+type usbmanagerContext struct {
+	agentbase.AgentBase
+	SubAssignableAdapters pubsub.Subscription
+	SubDomainStatus       pubsub.Subscription
+	controller            usbmanagerController
+	runningDomains        map[string]struct{} // mapping DomainStatus.DomainName
+}
+
+func newUsbmanagerContext() *usbmanagerContext {
+	usbCtx := &usbmanagerContext{}
+	usbCtx.controller.init()
+
+	usbCtx.runningDomains = make(map[string]struct{})
+
+	return usbCtx
+}
+
+// Run - Main function - invoked from zedbox.go
+func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, arguments []string) int {
+	logger = loggerArg
+	log = logArg
+
+	usbCtx := newUsbmanagerContext()
+
+	agentbase.Init(usbCtx, logger, log, agentName,
+		agentbase.WithPidFile(),
+		agentbase.WithWatchdog(ps, warningTime, errorTime),
+		agentbase.WithArguments(arguments))
+
+	// Wait until we have been onboarded aka know our own UUID, but we don't use the UUID
+	err := utils.WaitForOnboarded(ps, log, agentName, warningTime, errorTime)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Functionf("processed onboarded")
+
+	if err := utils.WaitForVault(ps, log, agentName, warningTime, errorTime); err != nil {
+		log.Fatal(err)
+	}
+	log.Functionf("processed Vault Status")
+
+	var subs []pubsub.Subscription
+
+	currentHypervisor := domainmgr.CurrentHypervisor()
+	_, ok := currentHypervisor.(hypervisor.KvmContext)
+	if ok {
+		usbCtx.subscribe(ps)
+		subs = pubsub.SubscriptionsByReflection(usbCtx)
```

Comment:
This is maybe a bit too fancy solution :)
Not sure what value does it add apart from few code lines saved at the cost of obfuscating the code for the reader.

Reply:
No, this solution is absolutely necessary, because beforehand it was a nightmare, every time I changed the subscriptions they would not work, because I forgot to change it in `process` as well.

Reply:
Now that the code has stabilized, is the use of reflection still justified?
The thing is that usually subscriptions are used with handlers, so it is necessary to have custom initialization for each anyway. There will be very few use-cases for this function outside of this one.

Reply:
OK, now I get what you mean.
I think that long term we might want to rethink pubsub APIs and somehow avoid the need to write separate case statement for each subscription (which generates many code lines, e.g. [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/zedagent/zedagent.go#L774-L814)). It would for example help if `pubsub.PubSub` would allow to list all submitted subscriptions (then you would not need to discover them using reflection), etc.
Maybe even your `process` function could be generalized into some pubsub helper, but that is for another PR...